File: Trees_Graphs/python/route.py
from GraphQueue import GraphQueue
from Graph import GraphNode


def has_route(node1, node2):
    queue = GraphQueue()
    node1.marked = True
    queue.enqueue(node1)

    while not queue.is_empty():
        node = queue.dequeue()

        for n in node.neighbours:
            if n == node2:
                return True
            elif not n.marked:
                n.marked = True
                queue.enqueue(n)

    return False


def main():
    node_a = GraphNode("A")
    node_b = GraphNode("B")
    node_c = GraphNode("C")
    node_d = GraphNode("D")
    node_e = GraphNode("E")
    node_f = GraphNode("F")

    node_a.update_neighbours([node_b, node_e, node_f])

    node_b.update_neighbours([node_e])

    node_c.update_neighbours([node_b])

    node_d.update_neighbours([node_c, node_e])

    node_f.update_neighbours([node_b])

    print(has_route(node_f, node_e))


if __name__ == '__main__':
    main()

File: Trees_Graphs/go/route.go
package main

import (
  "types"
  "fmt"
)


func hasRoute(node1 *types.GraphNode, node2 *types.GraphNode) bool {
  var q types.GraphQueue
  node1.Marked = true
  q.Enqueue(node1)

  for !q.IsEmpty() {
    node, _ := q.Dequeue()

    for i := 0; i < len(node.Neighbours); i++ {
      n := node.Neighbours[i]

      if n == node2 {
        return true
      } else if !n.Marked {
        n.Marked = true
        q.Enqueue(n)
      }
    }
  }

  return false
}


func main() {
  nodeA := &types.GraphNode{Value: "A"}
  nodeB := &types.GraphNode{Value: "B"}
  nodeC := &types.GraphNode{Value: "C"}
  nodeD := &types.GraphNode{Value: "D"}
  nodeE := &types.GraphNode{Value: "E"}
  nodeF := &types.GraphNode{Value: "F"}

  nodeA.Neighbours = append(nodeA.Neighbours, nodeB)
  nodeA.Neighbours = append(nodeA.Neighbours, nodeE)
  nodeA.Neighbours = append(nodeA.Neighbours, nodeF)

  nodeB.Neighbours = append(nodeB.Neighbours, nodeE)

  nodeC.Neighbours = append(nodeC.Neighbours, nodeB)

  nodeD.Neighbours = append(nodeD.Neighbours, nodeC)
  nodeD.Neighbours = append(nodeD.Neighbours, nodeE)

  nodeF.Neighbours = append(nodeF.Neighbours, nodeB)

  fmt.Println(hasRoute(nodeA, nodeC))
}
